File: nim/day05.nim
import strutils

const instructions = readFile("./inputs/05.txt").splitLines.map(parseInt)


func stepByStep(originalInstructions: seq[int], secondPart = false): uint32 =
  var
    instructions = originalInstructions
    line, jump: int

  while line < instructions.len:
    jump = instructions[line]
    if secondPart and jump >= 3:
      dec instructions[line]
    else:
      inc instructions[line]
    line += jump
    inc result


echo instructions.stepByStep()
echo instructions.stepByStep(secondPart=true)

File: python/day05.py
with open('./inputs/05.txt') as f:
    original_instructions = [int(line) for line in f.readlines()]


def step_by_step(instructions, second_part=False):
    line = 0
    steps = 0

    while True:
        try:
            # EAFP (https://docs.python.org/3/glossary.html#term-eafp)
            # a bit faster than LBYL (repeteately checking if inside of bounds)
            jump = instructions[line]
        except IndexError:
            return steps

        if second_part and jump >= 3:
            instructions[line] -= 1
        else:
            instructions[line] += 1
        line += jump
        steps += 1


first = step_by_step(original_instructions.copy())
second = step_by_step(original_instructions.copy(), second_part=True)

print(first)
print(second)
